fix(avs): resolve more audit findings

trestinlsd opened the pull request:
## Description

<!-- Add a description of the changes that this PR introduces and the files that
are the most critical to review.
-->

----

1.add missing events
2.add missing pre compiled interfaces
3.fix audit TOW-02


<!--
< < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >

**All** items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.

PR review checkboxes:

I have...

- [ ] added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] included the correct
      [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json)
      in the PR title
- [ ] targeted the correct branch
      (see [PR Targeting](https://github.com/ExocoreNetwork/exocore/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link in the PR description to the relevant issue or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all required CI checks have passed

Code maintenance:

I have...

- [ ] written unit and integration [tests](https://github.com/ExocoreNetwork/exocore/blob/main/CONTRIBUTING.md#testing)
- [ ] added relevant [`godoc`](https://go.dev/blog/godoc) and [code comments](https://blog.jbowen.dev/2019/09/the-magic-of-go-comments/).
- [ ] updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)

______

### Reviewers Checklist

**All** items are required.
Please add a note if the item is not applicable
and please add your handle next to the items reviewed
if you only reviewed selected items.

I have...

- [ ] confirmed the correct
      [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json)
      in the PR title
- [ ] confirmed all author checklist items have been addressed
- [ ] confirmed that this PR does not change production code

-->

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Introduced multiple new events for AVS operations, including `AVSRegistered`, `AVSUpdated`, and `TaskSubmittedByOperator`.
	- New functions added for querying AVS information, such as `getAVSInfo`, `getTaskInfo`, and `isOperator`.
	- Operators can now submit task results through the new `operatorSubmitTask` function.
	- Transitioned to a two-phase task submission model with updated phase handling.
	- Added a new `Phase` enum to enhance clarity in task submission phases.

- **Bug Fixes**
	- Enhanced error handling across various methods for improved clarity and consistency.
	- Updated validation criteria for AVS parameters to reflect new phase requirements.

- **Documentation**
	- Updated comments and error messages for better understanding and clarity regarding AVS operations and parameters.

- **Refactor**
	- Streamlined event handling and method signatures for better organization and readability.
	- Improved address handling and consistency across various methods and structures.
	- Normalized address handling to be case-insensitive throughout the AVS operations.

- **Tests**
	- Expanded test coverage for new and modified functionalities, ensuring robust validation of AVS operations.
	- Updated tests to reflect changes in parameter types and validation logic.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

trestinlsd committed:
fix audit TOW-02:avs deregister

trestinlsd committed:
fix audit TOW-02:avs deregister

trestinlsd committed:
add some new precompile func

trestinlsd committed:
fix avs precompile test err

trestinlsd committed:
add events for all transactions

trestinlsd committed:
fix lint err

trestinlsd committed:
add precompile query func

trestinlsd committed:
Merge branch 'ExocoreNetwork:develop' into develop

trestinlsd commented:
@coderabbitai review this pr

MaxMustermann2 commented on x/avs/keeper/avs_test.go:
```diff
@@ -105,7 +105,7 @@ func (suite *AVSTestSuite) TestUpdateAVSInfo_DeRegister() {
 		AssetID:           assetID,
 		MinSelfDelegation: uint64(10),
-		UnbondingPeriod:   uint64(7),
+		UnbondingPeriod:   uint64(2),
```
Same comment.

trestinlsd replied:
fixed

MaxMustermann2 commented on precompiles/avs/IAVSManager.sol:
```diff
@@ -13,6 +13,40 @@ IAVSManager constant AVSMANAGER_CONTRACT = IAVSManager(
+        string[] signedOperators;
+        string[] noSignedOperators;
+        string[] errSignedOperators;
+        string taskTotalPower;
```
This has too many members and will probably error in Solidity.

trestinlsd replied:
fixed

MaxMustermann2 commented on precompiles/avs/avs.go:
```diff
@@ -175,7 +194,8 @@ func (p Precompile) Run(evm *vm.EVM, contract *vm.Contract, readOnly bool) (bz [
+		MethodDeregisterOperatorFromAVS, MethodCreateAVSTask, MethodRegisterBLSPublicKey, MethodChallenge,
+		MethodRegisterOperatorToExocore, MethodOperatorSubmitTask:
 		return true
 	case MethodGetRegisteredPubkey, MethodGetOptinOperators, MethodGetAVSUSDValue, MethodGetOperatorOptedUSDValue:
```
`MethodIsOperator`

MaxMustermann2 commented on precompiles/avs/avs_test.go:
```diff
@@ -193,7 +197,14 @@ func (suite *AVSManagerPrecompileSuite) TestRegisterAVS() {
 			// Run precompiled contract
 			bz, err := suite.precompile.Run(evm, contract, tc.readOnly)
-
+			suite.CommitAfter(time.Hour*24 + time.Nanosecond)
+			suite.CommitAfter(time.Hour*24 + time.Nanosecond)
+			suite.CommitAfter(time.Hour*24 + time.Nanosecond)
+			suite.CommitAfter(time.Hour*24 + time.Nanosecond)
+			suite.CommitAfter(time.Hour*24 + time.Nanosecond)
+			suite.CommitAfter(time.Hour*24 + time.Nanosecond)
+			suite.CommitAfter(time.Hour*24 + time.Nanosecond)
+			suite.CommitAfter(time.Hour*24 + time.Nanosecond)
```
Same comment - and instead of using a hardcoded number, please use a common constant which will be used during AVS registration as well as deregistration.

MaxMustermann2 commented on precompiles/avs/avs_test.go:
```diff
@@ -717,7 +728,7 @@
 		return suite.Address, input
 	}
-	successRet, err := suite.precompile.Methods[avs.MethodCreateAVSTask].Outputs.Pack(true)
+	successRet, err := suite.precompile.Methods[avs.MethodCreateAVSTask].Outputs.Pack(true, uint64(1))
```
Each one is returning different values. This comment is inaccurate.

MaxMustermann2 commented on precompiles/avs/avs_test.go:
```diff
@@ -806,3 +817,210 @@ func (suite *AVSManagerPrecompileSuite) TestRunRegTaskInfo() {
+	}
+}
+
+//func (suite *AVSManagerPrecompileSuite) TestOperatorSubmitTask() {
```
Will this be added back?

MaxMustermann2 commented on x/avs/keeper/avs_test.go:
```diff
@@ -119,7 +119,10 @@ func (suite *AVSTestSuite) TestUpdateAVSInfo_DeRegister() {
+	suite.CommitAfter(48*time.Hour + time.Nanosecond)
+	suite.CommitAfter(48*time.Hour + time.Nanosecond)
+	suite.CommitAfter(48*time.Hour + time.Nanosecond)
+	suite.CommitAfter(48*time.Hour + time.Nanosecond)
```
Same as before.

MaxMustermann2 commented on x/avs/keeper/params.go:
```diff
@@ -101,3 +101,21 @@ type ChallengeParams struct {
 	CallerAddress       string         `json:"caller_address"`
 }
+
+type TaskResultParams struct {
```
Instead of making these `struct`s in `keeper`, why not add them in `types` ?

MaxMustermann2 commented on precompiles/avs/tx.go:
```diff
@@ -305,5 +326,108 @@ func (p Precompile) RegisterBLSPublicKey(
+}
+
+// RegisterOperatorToExocore operator register.
+func (p Precompile) RegisterOperatorToExocore(
```
Is it necessary to add this function here? Operators should ideally register using the Cosmos transaction type only and not through this method.

Even if this function is added here, the commission, the earnings address, the approve address and the client chain address members of the `OperatorInfo` struct are not set or validated in this call.

CC @TimmyExogenous

trestinlsd replied:
fixed

MaxMustermann2 commented on x/avs/types/expected_keepers.go:
```diff
@@ -50,4 +54,5 @@ type AssetsKeeper interface {
 		ctx sdk.Context, assetID string,
 	) (info *assetstype.StakingAssetInfo, err error)
 	IsStakingAsset(sdk.Context, string) bool
+	QueOperatorAssetInfos(ctx context.Context, infos *assetstype.QueryOperatorAssetInfos) (*assetstype.QueryOperatorAssetInfosResponse, error)
```
Why not use `GetOperatorAssetInfos` directly instead of going via the query server?

trestinlsd replied:
fixed

trestinlsd replied:
QueOperatorAssetInfos is not actually used and has been removed

MaxMustermann2 commented on x/avs/types/expected_keepers.go:
```diff
@@ -42,6 +44,8 @@
 	GetOptedInOperatorListByAVS(ctx sdk.Context, avsAddr string) ([]string, error)
 	GetOperatorOptedUSDValue(ctx sdk.Context, avsAddr, operatorAddr string) (operatortypes.OperatorOptedUSDValue, error)
 	GetAVSUSDValue(ctx sdk.Context, avsAddr string) (sdkmath.LegacyDec, error)
+	SetOperatorInfo(ctx sdk.Context, addr string, info *operatortypes.OperatorInfo) (err error)
+	QueryOperatorInfo(ctx context.Context, req *operatortypes.GetOperatorInfoReq) (*operatortypes.OperatorInfo, error)
```
No, it's an interface. You can add more functions to what it exposes. Invalid comment.

MaxMustermann2 commented on precompiles/avs/query.go:
```diff
@@ -106,3 +113,85 @@
 	return method.Outputs.Pack(amount.ActiveUSDValue.BigInt())
 }
+
+func (p Precompile) GetAVSInfo(
+	ctx sdk.Context,
+	_ *vm.Contract,
+	method *abi.Method,
+	args []interface{},
+) ([]byte, error) {
+	if len(args) != len(p.ABI.Methods[MethodGetAVSInfo].Inputs) {
+		return nil, fmt.Errorf(cmn.ErrInvalidNumberOfArgs, len(p.ABI.Methods[MethodGetAVSInfo].Inputs), len(args))
+	}
+	// the key is set using the operator's acc address so the same logic should apply here
+	addr, ok := args[0].(common.Address)
+	if !ok {
+		return nil, fmt.Errorf(exocmn.ErrContractInputParaOrType, 0, "common.Address", addr)
+	}
+
+	avs, err := p.avsKeeper.QueryAVSInfo(ctx, &avstype.QueryAVSInfoReq{AVSAddress: addr.String()})
+	if err != nil {
+		// if the avs does not exist, return empty array
+		if strings.Contains(err.Error(), avstype.ErrNoKeyInTheStore.Error()) {
+			return method.Outputs.Pack("")
+		}
+		return nil, err
+	}
+
+	return method.Outputs.Pack(avs.GetInfo().EpochIdentifier)
+}
```
Also consider `errors.Cause` if needed.

MaxMustermann2 commented on precompiles/avs/avs.go:
```diff
@@ -133,8 +133,13 @@ func (p Precompile) Run(evm *vm.EVM, contract *vm.Contract, readOnly bool) (bz [
 		bz, err = p.GetOptedInOperatorAccAddrs(ctx, contract, method, args)
-	case MethodGetRegisteredPubkey:
-		bz, err = p.GetRegisteredPubkey(ctx, contract, method, args)
+	case MethodGetAVSInfo:
+		bz, err = p.GetAVSInfo(ctx, contract, method, args)
+	case MethodGetTaskInfo:
+		bz, err = p.GetTaskInfo(ctx, contract, method, args)
+	case MethodIsOperator:
+		bz, err = p.IsOperator(ctx, contract, method, args)
+
```
+1

MaxMustermann2 commented on precompiles/avs/events.go:
```diff
@@ -1,22 +1,328 @@
+	stateDB.AddLog(&ethtypes.Log{
+		Address:     p.Address(),
+		Topics:      topics,
+		Data:        packed,
+		BlockNumber: uint64(ctx.BlockHeight()),
+	})
+
+	return nil
+}
+
+func (p Precompile) EmitOperatorRegisteredToExocore(ctx sdk.Context, stateDB vm.StateDB, params *avskeep.OperatorParams) error {
+	// Prepare the event topics
+	event := p.ABI.Events[EventTypeOperatorRegisteredToExocore]
+
+	topics := make([]common.Hash, 1)
+
+	// The first topic is always the signature of the event.
+	topics[0] = event.ID
+
+	var err error
+
+	// Pack the arguments to be used as the Data field
+	arguments := event.Inputs[0:3]
+	packed, err := arguments.Pack(
+		common.HexToAddress(params.CallerAddress),
+		params.OperatorMetaInfo,
+		true)
+	if err != nil {
+		return err
+	}
+
+	stateDB.AddLog(&ethtypes.Log{
+		Address:     p.Address(),
+		Topics:      topics,
+		Data:        packed,
+		BlockNumber: uint64(ctx.BlockHeight()),
+	})
+
+	return nil
+}
```
+1

mikebraver replied:
@trestinlsd please follow this simplified comments for refactoring

MaxMustermann2 commented on precompiles/avs/IAVSManager.sol:
```diff
@@ -143,15 +177,39 @@ interface IAVSManager {
+        string memory metaInfo
+    ) external returns (bool success);
 
+    /// @dev operatorSubmitTask ,  this function enables a operator submit a task result.
+    /// @param sender The external address for calling this method.
+    /// @param taskID The id of task.
+    /// @param taskResponse is the task response data..
+    /// @param blsSignature is the operator bls sig info..
+    /// @param taskContractAddress is contract address of task.
+    /// @param stage this field is used to solve the problem of task results being copied by other operators.
+    //  It is a two-stage submission with two values, 1 and 2
+    function operatorSubmitTask(
+        address sender,
+        uint64 taskID,
+        bytes calldata taskResponse,
+        bytes calldata blsSignature,
+        address taskContractAddress,
+        string memory stage
+    ) external returns (bool success);
```
Good point

MaxMustermann2 commented on x/avs/types/expected_keepers.go:
```diff
@@ -42,6 +44,8 @@ type OperatorKeeper interface {
 	GetOperatorOptedUSDValue(ctx sdk.Context, avsAddr, operatorAddr string) (operatortypes.OperatorOptedUSDValue, error)
 	GetAVSUSDValue(ctx sdk.Context, avsAddr string) (sdkmath.LegacyDec, error)
+	SetOperatorInfo(ctx sdk.Context, addr string, info *operatortypes.OperatorInfo) (err error)
+	QueryOperatorInfo(ctx context.Context, req *operatortypes.GetOperatorInfoReq) (*operatortypes.OperatorInfo, error)
```
Similar question here: why not use `keeper.OperatorInfo(` directly?

trestinlsd replied:
fixed

mikebraver commented on precompiles/avs/IAVSManager.sol:
```diff
@@ -169,14 +227,24 @@ interface IAVSManager {
+    /// @param taskID The id of task.
+    function getTaskInfo(
+        address taskAddr,
+        uint64 taskID
```
Isn't it enough using just the `taskAddr`?

trestinlsd replied:
The task primary key is a unique combination of address and ID

trestinlsd committed:
Merge branch 'ExocoreNetwork:develop' into develop

trestinlsd committed:
obtain the specified current epoch based on epochIdentifier

trestinlsd commented on x/avs/keeper/params.go:
```diff
@@ -101,3 +101,21 @@ type ChallengeParams struct {
+	CallerAddress       string         `json:"caller_address"`
+}
+
+type OperatorParams struct {
+	EarningsAddr     string `json:"earnings_addr"`
+	ApproveAddr      string `son:"approve_addr"`
+	OperatorMetaInfo string `json:"operator_meta_info"`
+	CallerAddress    string `json:"caller_address"`
+}
```
fixed

mikebraver commented on precompiles/avs/events.go:
```diff
@@ -1,22 +1,328 @@
+	// Pack the arguments to be used as the Data field
+	arguments := event.Inputs[0:5]
+	packed, err := arguments.Pack(
+		common.HexToAddress(params.CallerAddress),
```
Seemed like the arguments number didn't match with the `event.Inputs[0:5]` amount

mikebraver commented on precompiles/avs/events.go:
```diff
@@ -1,22 +1,328 @@
+	packed, err := arguments.Pack(
+		common.HexToAddress(params.Operator),
+		params.Name,
+		true)
```
same here

mikebraver commented on precompiles/avs/events.go:
```diff
@@ -1,22 +1,328 @@
+	arguments := event.Inputs[0:3]
+	packed, err := arguments.Pack(
+		common.HexToAddress(params.CallerAddress),
+		params.OperatorMetaInfo,
```
check all the places

mikebraver commented on precompiles/avs/IAVSManager.sol:
```diff
@@ -13,6 +13,40 @@ IAVSManager constant AVSMANAGER_CONTRACT = IAVSManager(
+        string blsSignature, string taskContractAddress, string stage, bool success);
+
+
+    struct TaskInfo {
+        string taskContractAddress;
+        string name;
+        bytes hash;
+        uint64 taskId;
+        uint64 taskResponsePeriod;
+        uint64 taskStatisticalPeriod;
+        uint64 taskChallengePeriod;
+        uint64 thresholdPercentage;
+        uint64 startingEpoch;
+        uint64 actualThreshold;
+        string[] optInOperators;
+        string[] signedOperators;
+        string[] noSignedOperators;
+        string[] errSignedOperators;
+        string taskTotalPower;
+        // OperatorActivePowerList is not defined here as it's not clear from the provided message
+    }
```
@trestinlsd optimize as array like this

trestinlsd commented on precompiles/avs/avs.go:
```diff
@@ -133,8 +133,29 @@
+			bz, err = method.Outputs.Pack(false)
+		}
+	case MethodGetAVSInfo:
+		bz, err = p.GetAVSInfo(ctx, contract, method, args)
+		if err != nil {
+			ctx.Logger().Error("internal error when calling avs precompile", "module", "avs precompile", "method", method.Name, "err", err)
+			bz, err = method.Outputs.Pack(false)
+		}
```
fixed

trestinlsd committed:
fix some comments

trestinlsd committed:
fix lint err

bwhour commented on x/avs/keeper/keeper.go:
```diff
@@ -220,32 +225,34 @@ func (k Keeper) UpdateAVSInfo(ctx sdk.Context, params *types.AVSRegisterOrDeregi
 	avsInfo := k.GetAVSInfoByTaskAddress(ctx, params.TaskContractAddress)
 	if avsInfo.AvsAddress == "" {
-		return errorsmod.Wrap(types.ErrUnregisterNonExistent, fmt.Sprintf("the taskaddr is :%s", params.TaskContractAddress))
+		return 0, errorsmod.Wrap(types.ErrUnregisterNonExistent, fmt.Sprintf("the taskaddr is :%s", params.TaskContractAddress))
```
can you replace the numbers  0 with a meaningful variable like InvalidTaskID=0, it will be better to understand the meaning.

bwhour commented on precompiles/avs/IAVSManager.sol:
```diff
@@ -143,15 +158,31 @@ interface IAVSManager {
+    /// @param taskResponse is the task response data..
+    /// @param blsSignature is the operator bls sig info..
+    /// @param taskContractAddress is contract address of task.
+    /// @param stage this field is used to solve the problem of task results being copied by other operators.
```
The enumeration is 0 and 1, and here you have the strings 1 and 2. This is easy to confuse. You can directly use uint8 enumeration values.
it's a good practice that If you can use the normal type in the contract, use the normal type, and try to use as little string as possible, which consumes more gas.

trestinlsd committed:
fix comments

bwhour commented on precompiles/avs/IAVSManager.sol:
```diff
@@ -169,14 +199,29 @@ interface IAVSManager {
+    /// @param avsAddr The address of the avs
+    function getAVSInfo(
+        address avsAddr
+    ) external view returns (string memory epochIdentifier);
```
this returns epochIdentifier, not like the comment and function name Avs info.

bwhour commented on precompiles/avs/events.go:
```diff
@@ -41,3 +40,88 @@ func (p Precompile) EmitCreateAVSTaskEvent(ctx sdk.Context, stateDB vm.StateDB,
+func (p Precompile) EmitOperatorJoined(ctx sdk.Context, stateDB vm.StateDB, params *avstypes.OperatorOptParams) error {
+	arguments := p.ABI.Events[EventTypeOperatorJoined].Inputs
+	return p.emitEvent(ctx, stateDB, EventTypeOperatorJoined, arguments,
+		common.HexToAddress(params.OperatorAddress))
```
this OperatorAddress should be cosmos AccAddress ， it shouldn't be  transformed with ethereum common.HexToAddress

bwhour commented on precompiles/avs/events.go:
```diff
@@ -41,3 +40,88 @@ func (p Precompile) EmitCreateAVSTaskEvent(ctx sdk.Context, stateDB vm.StateDB,
+func (p Precompile) EmitOperatorOuted(ctx sdk.Context, stateDB vm.StateDB, params *avstypes.OperatorOptParams) error {
+	arguments := p.ABI.Events[EventTypeOperatorOuted].Inputs
+	return p.emitEvent(ctx, stateDB, EventTypeOperatorOuted, arguments,
+		common.HexToAddress(params.OperatorAddress))
```
here too： this OperatorAddress should be cosmos AccAddress ， it shouldn't be  transformed with ethereum common.HexToAddress

bwhour commented on precompiles/avs/events.go:
```diff
@@ -41,3 +40,88 @@ func (p Precompile) EmitCreateAVSTaskEvent(ctx sdk.Context, stateDB vm.StateDB,
+func (p Precompile) EmitPublicKeyRegistered(ctx sdk.Context, stateDB vm.StateDB, params *avstypes.BlsParams) error {
+	arguments := p.ABI.Events[EventTypePublicKeyRegistered].Inputs
+	return p.emitEvent(ctx, stateDB, EventTypePublicKeyRegistered, arguments,
+		common.HexToAddress(params.Operator),
```
is this operator address like ethereum ？

bwhour commented on precompiles/avs/avs_test.go:
```diff
@@ -210,18 +207,54 @@ func (suite *AVSManagerPrecompileSuite) TestRegisterAVS() {
+			MinStakeAmount:      minStakeAmount,
+			MinOptInOperators:   params[0],
+			MinTotalStakeAmount: params[1],
+			AvsReward:           sdk.MustNewDecFromStr(strconv.FormatUint(params[1], 10)),
```
consider using **strconv.Itoa(int(xxx))** instead of calling _strconv.FormatUint(xxx, 10)_, as it is more concise and performs better.

bwhour commented on precompiles/avs/query.go:
```diff
@@ -106,3 +114,106 @@ func (p Precompile) GetOperatorOptedUSDValue(
+	if len(args) != len(p.ABI.Methods[MethodGetAVSInfo].Inputs) {
+		return nil, fmt.Errorf(cmn.ErrInvalidNumberOfArgs, len(p.ABI.Methods[MethodGetAVSInfo].Inputs), len(args))
+	}
+	// the key is set using the operator's acc address so the same logic should apply here
```
the comment seems useless, is it operator address ?

bwhour commented on precompiles/avs/query.go:
```diff
@@ -106,3 +114,106 @@ func (p Precompile) GetOperatorOptedUSDValue(
+	if len(args) != len(p.ABI.Methods[MethodGetTaskInfo].Inputs) {
+		return nil, fmt.Errorf(cmn.ErrInvalidNumberOfArgs, len(p.ABI.Methods[MethodGetTaskInfo].Inputs), len(args))
+	}
+	// the key is set using the operator's acc address so the same logic should apply here
```
here the comment

trestinlsd committed:
fix comments

trestinlsd committed:
update for comments

bwhour commented on precompiles/avs/types.go:
```diff
@@ -228,7 +228,7 @@ func (p Precompile) GetTaskParamsFromInputs(_ sdk.Context, args []interface{}) (
 	if !ok || (callerAddress == common.Address{}) {
```
Given the CallerAddress is sdk.AccAddress,  so the type assertion  to common.Address is not reasonable. plase check all such code and the return error : 

> callerAddress, ok := args[0].(common.Address)

bwhour reviewed (approved):
it looks good, pls also look into the  reports from the ai-bot

MaxMustermann2 commented on precompiles/avs/IAVSManager.sol:
```diff
@@ -7,12 +7,26 @@ address constant AVSMANAGER_PRECOMPILE_ADDRESS = 0x00000000000000000000000000000
 IAVSManager constant AVSMANAGER_CONTRACT = IAVSManager(
```
Please `forge fmt` this file with `exocore-contracts/foundry.toml` configuration.

trestinlsd replied:
fixed

MaxMustermann2 commented on go.mod:
```diff
@@ -271,8 +271,10 @@ replace (
 	github.com/evmos/evmos/v16 => github.com/ExocoreNetwork/evmos/v16 v16.0.3-0.20240828081344-d5cfcd34a812
```
Are the changes in this file (ledgerwatch/interfaces and websocket) a result of a security advisory?

trestinlsd replied:
fixed

MaxMustermann2 commented on precompiles/avs/IAVSManager.sol:
```diff
@@ -7,12 +7,26 @@ address constant AVSMANAGER_PRECOMPILE_ADDRESS = 0x00000000000000000000000000000
+    event AVSDeregistered(address indexed avsAddr, string sender, string avsName);
+    event OperatorJoined(address indexed avsAddr,string sender);
+    event OperatorOuted(address indexed avsAddr,string sender);
+    event TaskCreated(string sender,uint64 taskId,address indexed taskContractAddress,string name,bytes hash,
+        uint64 taskResponsePeriod,uint64 taskChallengePeriod,uint64 thresholdPercentage,uint64 taskStatisticalPeriod
+    );
```
It makes sense to make this edit, definitely points 2 and 3.

Regarding point 1, the `sender` is an `exo1....` address. However, within our chain, each address can be represented as a hex address **AND** a bech32 encoded address. In Solidity, the former type is well defined and can be used well, however, the latter is not easy to use (for example, even checking whether two strings are equal is not directly permitted in Solidity). It would make sense to use the hex version, if possible, within Solidity and the bech32 version elsewhere. Given that these are events which Solidity cannot access, I don't see any reasons to switch to hex here. Additionally, `indexed` keyword will hash the string leaving the original irrecoverable. What do you think @coderabbitai ?

trestinlsd replied:
Given that these events are intended for external consumption and that Solidity does not interact with them directly, it makes sense to retain the sender as a string without the indexed keyword.

MaxMustermann2 replied:
What about points 2 and 3 @trestinlsd ? Do you agree or disagree with them ?

MaxMustermann2 replied:
Discussed that these are not super relevant and the number of members in the event does not surpass Solidity limits. Closing.

MaxMustermann2 commented on precompiles/avs/IAVSManager.sol:
```diff
@@ -7,12 +7,26 @@
 /// @dev The interface through which solidity contracts will interact with AVS-Manager
 /// @custom:address 0x0000000000000000000000000000000000000901
 interface IAVSManager {
+    event AVSRegistered(address indexed avsAddr, string sender, string avsName);
+    event AVSUpdated(address indexed avsAddr, string sender, string avsName);
+    event AVSDeregistered(address indexed avsAddr, string sender, string avsName);
+    event OperatorJoined(address indexed avsAddr,string sender);
+    event OperatorOuted(address indexed avsAddr,string sender);
+    event TaskCreated(string sender,uint64 taskId,address indexed taskContractAddress,string name,bytes hash,
+        uint64 taskResponsePeriod,uint64 taskChallengePeriod,uint64 thresholdPercentage,uint64 taskStatisticalPeriod
+    );
+    event ChallengeInitiated(string sender, bytes taskHash, uint64 taskID, bytes taskResponseHash,
+        string operatorAddress);
+    event PublicKeyRegistered(string sender, string name);
+    event TaskSubmittedByOperator(string sender, uint64 taskID, bytes taskResponse,
+        bytes blsSignature, address indexed taskContractAddress, uint8 phase);
+
```
+1 for the formatting changes, but most of them can be done via `forge fmt`

MaxMustermann2 commented on precompiles/avs/IAVSManager.sol:
```diff
@@ -169,14 +199,29 @@ interface IAVSManager {
-        uint64 thresholdPercentage,
-        uint64 taskStatisticalPeriod
-    );
+    /// @dev getAVSInfo  is a function to query Avs epochIdentifier info.
+    /// @param avsAddr The address of the avs
+    function getAVSInfo(
+        address avsAddr
+    ) external view returns (string memory epochIdentifier);
```
+1

MaxMustermann2 commented on precompiles/avs/IAVSManager.sol:
```diff
@@ -169,14 +199,29 @@
+        address avsAddr
+    ) external view returns (string memory epochIdentifier);
+
+    /// @dev getTaskInfo  is a function to query task info.
+    /// @param taskAddr The address of the avs task
+    /// @param taskID The id of task.
+    function getTaskInfo(
+        address taskAddr,
+        uint64 taskID
+    ) external view returns (uint64[] memory info);
```
+1

MaxMustermann2 commented on precompiles/avs/IAVSManager.sol:
```diff
@@ -143,15 +157,32 @@ interface IAVSManager {
     /// @dev Returns the operators of all opt-in in the current avs
     /// @param avsAddress avs address
-    function getOptInOperators(address avsAddress) external returns (string[] memory operators);
+    function getOptInOperators(address avsAddress) external view returns (string[] memory operators);
```
Should this be `string[]` or `address[]` ? The former looks better to users while the latter is better suited for Solidity operations.

trestinlsd replied:
The returned address data starts with exo, so it is better to use a string array

MaxMustermann2 replied:
Each `exo1` address can be converted to `0x` address by using `sdk.AccAddress(hexAddress.Bytes())` and vice-versa by using `common.Address(accAddress.Bytes())`. Solidity has much better support for manipulating the `0x` addresses while it cannot do much with the `exo1` string addresses.

Given that this function is [not just events intended for external consumption](https://github.com/ExocoreNetwork/exocore/pull/220#discussion_r1820021539), I think it would make sense for the function to return the well-supported native Solidity `address` type. Outside of Solidity, it can be converted to either format easily; however, such conversions are not possible within Solidity.

CC @adu-web3: any thoughts?

adu-web3 replied:
yeah, `exo1` prefix is added when a normal evm address is encoded to its bech32 representation. So how would we use these operator addresses? most of our precompile functions treat operator address as hex string and assume it would be passed from some EOA caller, so it might be consistent with these functions if we assume the operator addresses returned would be passed into these functions as arguments

MaxMustermann2 replied:
Closing for now. Agree that the bech32-encoded address is more consistent with the rest of the system; however, it would perhaps be a good idea to keep an open mind about changing this later.

MaxMustermann2 commented on precompiles/avs/IAVSManager.sol:
```diff
@@ -7,12 +7,26 @@ address constant AVSMANAGER_PRECOMPILE_ADDRESS = 0x00000000000000000000000000000
+    event AVSUpdated(address indexed avsAddr, string sender, string avsName);
+    event AVSDeregistered(address indexed avsAddr, string sender, string avsName);
+    event OperatorJoined(address indexed avsAddr,string sender);
+    event OperatorOuted(address indexed avsAddr,string sender);
```
```suggestion
    event OperatorLeft(address indexed avsAddr,string sender);
```

trestinlsd replied:
fixed

MaxMustermann2 commented on precompiles/avs/query.go:
```diff
@@ -102,7 +115,111 @@ func (p Precompile) GetOperatorOptedUSDValue(
+		return nil, fmt.Errorf(exocmn.ErrContractInputParaOrType, 0, "common.Address", addr)
+	}
+
+	avs, err := p.avsKeeper.QueryAVSInfo(ctx, &avstype.QueryAVSInfoReq{AVSAddress: addr.String()})
```
Why do you use the `QueryAVSInfo` interface instead of `GetAVSInfo` directly?

trestinlsd replied:
fixed

MaxMustermann2 commented on precompiles/avs/query.go:
```diff
@@ -102,7 +115,111 @@ func (p Precompile) GetOperatorOptedUSDValue(
+		return nil, fmt.Errorf(exocmn.ErrContractInputParaOrType, 1, "uint64", taskID)
+	}
+
+	task, err := p.avsKeeper.QueryAVSTaskInfo(ctx, &avstype.QueryAVSTaskInfoReq{TaskAddr: addr.String(), TaskId: strconv.FormatUint(taskID, 10)})
```
Why not use `GetTaskInfo` ?

trestinlsd replied:
fixed

MaxMustermann2 commented on precompiles/avs/tx.go:
```diff
@@ -305,5 +322,79 @@ func (p Precompile) RegisterBLSPublicKey(
+	}
+	resultParams.Phase = phase
+
+	resultParams.OperatorAddress = resultParams.CallerAddress
```
+1, why do we need two different fields?

MaxMustermann2 commented on precompiles/avs/tx.go:
```diff
@@ -305,5 +322,79 @@ func (p Precompile) RegisterBLSPublicKey(
+	resultParams.TaskContractAddress = taskAddr
+
+	phase, ok := args[5].(uint8)
+	if !ok {
```
No validation of the `phase` number? Only certain values are supported right?

trestinlsd replied:
fixed

MaxMustermann2 commented on precompiles/avs/types.go:
```diff
@@ -126,7 +125,7 @@ func (p Precompile) GetAVSParamsFromUpdateInputs(_ sdk.Context, args []interface
 		return nil, fmt.Errorf(exocmn.ErrContractInputParaOrType, 0, "common.Address", callerAddress)
 	}
-	avsParams.CallerAddress = sdk.AccAddress(callerAddress[:]).String()
+	avsParams.CallerAddress = callerAddress[:]
```
This is already done. Check again @coderabbitai

MaxMustermann2 commented on x/avs/keeper/avs_test.go:
```diff
@@ -117,40 +118,41 @@ func (suite *AVSTestSuite) TestUpdateAVSInfo_DeRegister() {
+	suite.CommitAfter(48*time.Hour + time.Nanosecond)
+	suite.CommitAfter(48*time.Hour + time.Nanosecond)
+	suite.CommitAfter(48*time.Hour + time.Nanosecond)
+	suite.CommitAfter(48*time.Hour + time.Nanosecond)
```
Do not type the same line many times. Instead, use a constant in the test and loop over it.

Currently, the unbonding period is set to 2 days. However, you are committing 48 hours * 4 = 8 days. Is that intentional? You should instead run the test after committing 2 days. That way the functionality can be verified.

MaxMustermann2 commented on x/avs/types/utils.go:
```diff
@@ -0,0 +1,94 @@
+	BlsSignature        []byte         `json:"bls_signature"`
+	TaskContractAddress common.Address `json:"task_contract_address"`
+	TaskID              uint64         `json:"task_id"`
+	Phase               uint8          `json:"phase"`
```
Consider making this an enum.

MaxMustermann2 commented on x/avs/types/utils.go:
```diff
@@ -0,0 +1,94 @@
+	TaskResponseHash      string         `json:"task_response_hash"`
+	TaskResponse          []byte         `json:"task_response"`
+	BlsSignature          []byte         `json:"bls_signature"`
+	Phase                 string         `json:"phase"`
```
Not clear, why is this a string?

trestinlsd replied:
fixed

trestinlsd replied:
This field 'phase' is redundant in the TaskInfoParams struct and has been removed

MaxMustermann2 commented on precompiles/avs/IAVSManager.sol:
```diff
@@ -4,15 +4,43 @@
+        string sender, bytes taskHash, uint64 taskID, bytes taskResponseHash, string operatorAddress
+    );
+    event PublicKeyRegistered(string sender, string name);
+    event TaskSubmittedByOperator(
+        string sender,
+        uint64 taskID,
+        bytes taskResponse,
+        bytes blsSignature,
+        address indexed taskContractAddress,
+        uint8 phase
+    );
```
@coderabbitai: Please refrain from such suggestions. We do not want to add the `indexed` keyword intentionally since it results in the information being hashed irretrievably.

MaxMustermann2 commented on precompiles/avs/IAVSManager.sol:
```diff
@@ -143,40 +162,66 @@ interface IAVSManager {
+
+    /// @dev isOperator checks if the given address is registered as an operator on exocore.
+    /// @param operatorAddr The address of the operator
+    function isOperator(string memory operatorAddr) external view returns (bool);
```
Could potentially consider a similar `function isOperator(address operatorAddr) external view returns (bool)`

MaxMustermann2 commented on precompiles/avs/foundry.toml:
```diff
@@ -0,0 +1,8 @@
+
```
Remove this file before merging, thanks.

MaxMustermann2 commented on precompiles/avs/query.go:
```diff
@@ -102,7 +115,111 @@ func (p Precompile) GetOperatorOptedUSDValue(
+	if !ok {
+		return nil, fmt.Errorf(exocmn.ErrContractInputParaOrType, 0, "common.Address", addr)
+	}
+	taskID, ok := args[1].(uint64)
+	if !ok {
+		return nil, fmt.Errorf(exocmn.ErrContractInputParaOrType, 1, "uint64", taskID)
+	}
```
We chose the `uint64` type intentionally. Do not suggest changing to `uint256` without reason.

MaxMustermann2 commented on precompiles/avs/query.go:
```diff
@@ -102,7 +115,111 @@ func (p Precompile) GetOperatorOptedUSDValue(
+	}
+	epoch, flag := p.avsKeeper.GetEpochKeeper().GetEpochInfo(ctx, epochIdentifier)
+	if !flag {
+		return nil, nil
```
Should return an error here: `nil, err`

MaxMustermann2 commented on precompiles/avs/query_test.go:
```diff
@@ -313,3 +318,233 @@
+
+func (suite *AVSManagerPrecompileSuite) TestGetRegisteredPubkey() {
+	method := suite.precompile.Methods[avsManagerPrecompile.MethodGetRegisteredPubkey]
+	privateKey, err := blst.RandKey()
```
False, we have `suite.NoError(err)` just below.

MaxMustermann2 commented on x/avs/keeper/avs.go:
```diff
@@ -85,7 +85,7 @@ func (k *Keeper) GetEpochEndAVSs(ctx sdk.Context, epochIdentifier string, ending
 func (k *Keeper) GetAVSInfoByTaskAddress(ctx sdk.Context, taskAddr string) types.AVSInfo {
 	var avs types.AVSInfo
-	if taskAddr == "" {
+	if taskAddr == "" || taskAddr == "0x0000000000000000000000000000000000000000" {
```
```suggestion
	if taskAddr == "" || taskAddr == (common.Address{}).String() {
```

MaxMustermann2 commented on x/avs/keeper/avs.go:
```diff
@@ -85,7 +85,7 @@ func (k *Keeper) GetEpochEndAVSs(ctx sdk.Context, epochIdentifier string, ending
 // TODO:this function is frequently used while its implementation iterates over existing avs to find the target avs by task contract address,  we should use a reverse mapping to avoid iteration
 func (k *Keeper) GetAVSInfoByTaskAddress(ctx sdk.Context, taskAddr string) types.AVSInfo {
 	var avs types.AVSInfo
```
Should this be initialized? Or maybe an error be returned if not found?

MaxMustermann2 commented on x/avs/keeper/keeper.go:
```diff
@@ -330,7 +322,7 @@ func (k Keeper) GetAVSInfo(ctx sdk.Context, addr string) (*types.QueryAVSInfoRes
 	value := store.Get(common.HexToAddress(addr).Bytes())
 	if value == nil {
-		return nil, errorsmod.Wrap(types.ErrNoKeyInTheStore, fmt.Sprintf("GetAVSInfo: key is %s", addr))
+		return nil, errorsmod.Wrap(types.ErrNoKeyInTheStore, fmt.Sprintf("getAVSEpochIdentifier: key is %s", addr))
```
Undo this change

MaxMustermann2 commented on proto/exocore/avs/v1/tx.proto:
```diff
@@ -226,9 +226,9 @@ message TaskResultInfo {
-  string stage = 7;
+  // phase this field is used to solve the problem of task results being copied by other operators.
+  // It is a two-phase submission with two values, 0 and 1
+  uint32 phase = 7;
```
Make it a `uint8` or `enum` ?

MaxMustermann2 commented on x/avs/types/utils.go:
```diff
@@ -0,0 +1,104 @@
+	Name            string         `json:"name"`
+	BlsPublicKey    string         `json:"bls_public_key"`
+	IsRegistered    bool           `json:"is_registered"`
+	Action          uint64         `json:"action"`
```
+1

trestinlsd committed:
fix comments of max

MaxMustermann2 committed:
fix(tests): use correct case

MaxMustermann2 committed:
fix: use blank slash contract not 0x0000

trestinlsd committed:
update proto enum phase

MaxMustermann2 committed:
revert unnecessary change

The `runTx` ensures that, upon a message failure, the cache is not
written to disk.

MaxMustermann2 commented on precompiles/avs/IAVSManager.sol:
```diff
@@ -143,40 +162,65 @@ interface IAVSManager {
+    /// @param taskResponse is the task response data..
+    /// @param blsSignature is the operator bls sig info..
+    /// @param taskContractAddress is contract address of task.
+    /// @param phase The phase of the Two-Phase Commit protocol:
+    ///             1 = Prepare phase (commit preparation)
+    ///             2 = Commit phase (final commitment)
```
It has been changed to 1 and 2. Please check the entire PR to verify that it has been actioned.

MaxMustermann2 commented on precompiles/avs/avs_test.go:
```diff
@@ -210,18 +207,54 @@ func (suite *AVSManagerPrecompileSuite) TestRegisterAVS() {
 
+		err := suite.App.AVSManagerKeeper.SetAVSInfo(suite.Ctx, avs)
+		suite.NoError(err)
+		for i := 0; i < int(avsUnbondingPeriod)+2; i++ {
```
Do you know why 2 is used here instead of 1?

MaxMustermann2 commented on precompiles/avs/tx.go:
```diff
@@ -305,5 +322,85 @@ func (p Precompile) RegisterBLSPublicKey(
+
+	// The phase of the Two-Phase Commit protocol:
+	// 0 = Prepare phase (commit preparation)
+	// 1 = Commit phase (final commitment)
```
Update the comment

MaxMustermann2 commented on precompiles/avs/tx.go:
```diff
@@ -305,5 +322,85 @@
+		return nil, fmt.Errorf(exocmn.ErrContractInputParaOrType, 3, "[]byte", blsSignature)
+	}
+	resultParams.BlsSignature = blsSignature
+
+	taskAddr, ok := args[4].(common.Address)
+	if !ok || (taskAddr == common.Address{}) {
+		return nil, fmt.Errorf(exocmn.ErrContractInputParaOrType, 4, "common.Address", taskAddr)
+	}
+	resultParams.TaskContractAddress = taskAddr
+
+	phase, ok := args[5].(uint8)
+	if !ok {
+		return nil, fmt.Errorf("invalid phase type: expected uint8, got %T", args[5])
+	}
+
+	// The phase of the Two-Phase Commit protocol:
+	// 0 = Prepare phase (commit preparation)
+	// 1 = Commit phase (final commitment)
+	// validation of the phase number
+	if err := avstypes.ValidatePhase(avstypes.CommitPhase(phase)); err != nil {
+		return nil, fmt.Errorf("invalid phase value: %d. Expected 1 (Prepare) or 2 (Commit)", phase)
+	}
+	resultParams.Phase = avstypes.CommitPhase(phase)
+
+	result := &avstypes.TaskResultInfo{
+		TaskId:              resultParams.TaskID,
+		OperatorAddress:     resultParams.CallerAddress.String(),
+		TaskContractAddress: resultParams.TaskContractAddress.String(),
+		TaskResponse:        resultParams.TaskResponse,
+		BlsSignature:        resultParams.BlsSignature,
+		Phase:               avstypes.Phase(resultParams.Phase),
+	}
+	err := p.avsKeeper.SetTaskResultInfo(ctx, resultParams.CallerAddress.String(), result)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := p.EmitTaskSubmittedByOperator(ctx, stateDB, resultParams); err != nil {
+		return nil, err
+	}
```
Phase value should be 1/2 not 0/1. Re-run the check.

MaxMustermann2 commented on x/avs/types/utils.go:
```diff
@@ -0,0 +1,107 @@
+)
+
+// CommitPhase represents the phases of the Two-Phase Commit protocol
+type CommitPhase uint8
```
Do we need both `type CommitPhase uint8` and the one defined in `tx.proto` as `enum Phase` ?

MaxMustermann2 committed:
refactor(avs): remove unnecessary CommitPhase

MaxMustermann2 commented on x/avs/types/utils.go:
```diff
@@ -0,0 +1,99 @@
+	BlsSignature        []byte         `json:"bls_signature"`
+	TaskContractAddress common.Address `json:"task_contract_address"`
+	TaskID              uint64         `json:"task_id"`
+	Phase               Phase          `json:"phase"`
```
It is defined in `tx.pb.go` via `tx.proto`

MaxMustermann2 commented on x/avs/types/utils.go:
```diff
@@ -0,0 +1,99 @@
+	ErrSignedCount        uint64         `json:"err_signed_count"`
+	CallerAddress         sdk.AccAddress `json:"caller_address"`
+}
+type BlsParams struct {
+	OperatorAddress               sdk.AccAddress
+	Name                          string
+	PubKey                        []byte
+	PubkeyRegistrationSignature   []byte
+	PubkeyRegistrationMessageHash []byte
+}
```
Internal usage only; not user-facing

MaxMustermann2 commented on proto/exocore/avs/v1/tx.proto:
```diff
@@ -206,12 +198,20 @@
   AVSInfo info = 2;
 }
-
+// It is a two-phase submission with two values, 1 and 2
+enum Phase {
+  option (gogoproto.goproto_enum_prefix) = false;
+  // Default value when phase is not specified
+  PHASE_UNSPECIFIED = 0 [(gogoproto.enumvalue_customname) = "PhaseUnspecified"];
+  // First phase where operators prepare and submit their initial responses
+  PHASE_PREPARE = 1 [(gogoproto.enumvalue_customname) = "PhasePrepare"];
+  // Second phase where operators commit their prepared responses
+  PHASE_DO_COMMIT = 2 [(gogoproto.enumvalue_customname) = "PhaseDoCommit"];
+}
```
0 is mandatory for proto so this is okay.

MaxMustermann2 commented on x/avs/types/utils.go:
```diff
@@ -0,0 +1,99 @@
+	PubkeyRegistrationMessageHash []byte
+}
+
+type ProofParams struct {
+	TaskID              string
+	TaskContractAddress common.Address
+	AvsAddress          common.Address
+	Aggregator          string
+	OperatorStatus      []OperatorStatusParams
+	CallerAddress       sdk.AccAddress
+}
```
Structure is currently unused. When fully implemented, this will be fixed.

trestinlsd committed:
format the address in lowercase

trestinlsd committed:
Fix the missing field of topic in the event

cloud8little commented:
Testing.

cloud8little commented:
The AVS interface in precompiles/avs/query.go needs to convert AVS addresses to lowercase

trestinlsd committed:
update hex address format

trestinlsd committed:
Merge branch 'ExocoreNetwork:develop' into develop

cloud8little commented:
Operator opt-in the new registered avs `0xa485396eea13d26c88818e9cfe857229cd82689d`, but `queryAVSUSDValue `and `queryOperatorUSDValue `failed.

```
 exocored q operator get-avs-list exo1v5zhaezzpgndmvexru4nkm3cpywghdntncjx8v
avs_list:
- 0xa485396eea13d26c88818e9cfe857229cd82689d
exocored q operator get-operator-list 0xa485396eea13d26c88818e9cfe857229cd82689d
operator_list:
- exo1v5zhaezzpgndmvexru4nkm3cpywghdntncjx8v
```

```
exocored q assets QueOperatorAssetInfos exo1v5zhaezzpgndmvexru4nkm3cpywghdntncjx8v
asset_infos:
- asset_id: 0xdac17f958d2ee523a2206206994597c13d831ec7_0x65
  info:
    operator_share: "10.000000000000000000"
    pending_undelegation_amount: "0"
    total_amount: "10"
    total_share: "10.000000000000000000"
```
```
exocored q operator QueryOperatorUSDValue exo1v5zhaezzpgndmvexru4nkm3cpywghdntncjx8v 0xa485396eea13d26c88818e9cfe857229cd82689d
Error: rpc error: code = Unknown desc = GetOperatorOptedUSDValue: key is 0xa485396eea13d26c88818e9cfe857229cd82689d/exo1v5zhaezzpgndmvexru4nkm3cpywghdntncjx8v: there is no such key in the store: unknown request

exocored q operator QueryAVSUSDValue 0xa485396eea13d26c88818e9cfe857229cd82689d
amount: "0.000000000000000000"
```

trestinlsd committed:
format the address in lowercase for operator module

trestinlsd committed:
fix type errors in ABI

cloud8little commented:
Test passed with 0c527a16f300661678acaea814607bb54eea305e, check all these events emission successfully.

```
  "event TaskCreated(address indexed,uint64 indexed,string,string,bytes,uint64,uint64,uint64,uint64)",
  "event AVSDeregistered(address indexed,string,string)",
  "event AVSRegistered(address indexed,string,string)",
  "event AVSUpdated(address indexed,string,string)",
  "event ChallengeInitiated(string,bytes,uint64,bytes,string)",
  "event OperatorJoined(address,string)",
  "event OperatorLeft(address,string)",
  "event PublicKeyRegistered(string,string)",
  "event TaskSubmittedByOperator(address indexed,uint64 indexed,string,bytes,bytes,uint8)",
```
